### PR TITLE
[feature] (#320): added block streaming support + test

### PR DESCRIPTION
**Pull request by gv-timur**
added block streaming support + test

### CODE REVIEW

File: modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/client/Iroha2Client.kt [hunk 1]
```diff
@@ -196,6 +201,39 @@ open class Iroha2Client(
...
+                path = WS_ENDPOINT_BLOCK_STREAM
+            ) {
+                logger.debug("WebSocket opened")
+                val request = VersionedBlockSubscriptionRequest.V1(
+                    BlockSubscriptionRequest(BigInteger.valueOf(startFromBlock))
+                )
+                val payload = VersionedBlockSubscriptionRequest.encode(request)
+                send(payload.toFrame())
+                for (frame in incoming) {
+                    logger.debug("Received frame: $frame")
```
Comment (arndey) [inline modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/client/Iroha2Client.kt:224]:
```suggestion
                    logger.debug("Received frame: {}", frame)
```

File: modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/client/Iroha2Client.kt [hunk 2]
```diff
@@ -196,6 +201,39 @@ open class Iroha2Client(
...
+                port = peerUrl.port,
+                path = WS_ENDPOINT_BLOCK_STREAM
+            ) {
+                logger.debug("WebSocket opened")
+                val request = VersionedBlockSubscriptionRequest.V1(
+                    BlockSubscriptionRequest(BigInteger.valueOf(startFromBlock))
+                )
+                val payload = VersionedBlockSubscriptionRequest.encode(request)
+                send(payload.toFrame())
+                for (frame in incoming) {
```
Comment (arndey) [inline modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/client/Iroha2Client.kt:223]: perfect place to use kotlin channels or flows

Comment (arndey) [reply modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/client/Iroha2Client.kt:223]: may need to get rid of the launch above

File: modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            .cast<Expression.Raw>().value
+            .cast<Value.Identifiable>()
+            .identifiableBox.cast<IdentifiableBox.NewDomain>()
+            .newDomain.id.name.string
+        assertEquals(DEFAULT_DOMAIN_ID.asString(), registerDomain)
+        assertEquals(ALICE_ACCOUNT_ID.asString(), instructions[1].cast<Instruction.Register>().extractAccount().id.asString())
+        assertEquals(BOB_ACCOUNT_ID.asString(), instructions[2].cast<Instruction.Register>().extractAccount().id.asString())
+        assertEquals("foo@wonderland", instructions[3].cast<Instruction.Register>().extractAccount().id.asString())
+
+        fun checkSecondBlock(blockMessage: VersionedBlockMessage) {
```
Comment (arndey) [inline modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt:64]: please don't create function in the middle of another function
if it's necessary to keep function in function - move it to the top

File: modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        var payload = getInstructionPayload(committedBlock)
+        var instructionAccountDomain = payload.accountId.domainId.name.string
+        var instructionAccount = payload.accountId.name.string
+        val genesis = "genesis"
+        assertEquals(genesis, instructionAccountDomain)
+        assertEquals(genesis, instructionAccount)
+
+        var instructions = payload.instructions.cast<Executable.Instructions>().vec
+        assertEquals(4, instructions.size)
+        val registerDomain = instructions[0].cast<Instruction.Register>()
```
Comment (arndey) [inline modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt:53]: i think we might have some extract function fo that
if not - please create it

File: modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            .registerBox.`object`.expression
+            .cast<Expression.Raw>().value
+            .cast<Value.Identifiable>()
+            .identifiableBox.cast<IdentifiableBox.NewDomain>()
+            .newDomain.id.name.string
+        assertEquals(DEFAULT_DOMAIN_ID.asString(), registerDomain)
+        assertEquals(ALICE_ACCOUNT_ID.asString(), instructions[1].cast<Instruction.Register>().extractAccount().id.asString())
+        assertEquals(BOB_ACCOUNT_ID.asString(), instructions[2].cast<Instruction.Register>().extractAccount().id.asString())
+        assertEquals("foo@wonderland", instructions[3].cast<Instruction.Register>().extractAccount().id.asString())
```
Comment (arndey) [inline modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt:62]: please use const DEFAULT_DOMAIN

File: modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+        fun checkSecondBlock(blockMessage: VersionedBlockMessage) {
+            committedBlock = getCommittedBlock(blockMessage)
+            height = committedBlock.header.height.toLong()
+            assertEquals(2, height)
+            payload = getInstructionPayload(committedBlock)
+            instructionAccountDomain = payload.accountId.domainId.name.string
+            instructionAccount = payload.accountId.name.string
+            assertEquals(defaultDomain, instructionAccountDomain)
+            assertEquals("bob", instructionAccount)
```
Comment (arndey) [inline modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt:72]: please use const BOB_ACCOUNT_NAME
if you need it as string - please create new const

File: modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            instructions = payload.instructions.cast<Executable.Instructions>().vec
+            assertEquals(1, instructions.size)
+            val newAssetDefinition = instructions[0].cast<Instruction.Register>()
```
Comment (arndey) [inline modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt:75]: use extract fun

File: modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            registerAssetDefinition(AssetDefinitionId(newAssetName.asName(), DEFAULT_DOMAIN_ID), AssetValueType.Store())
+        }
+        var blocks = blocksResult.await()
+        var committedBlock = getCommittedBlock(blocks[0])
+        var height = committedBlock.header.height.toLong()
+        assertEquals(1, height)
+        var payload = getInstructionPayload(committedBlock)
+        var instructionAccountDomain = payload.accountId.domainId.name.string
+        var instructionAccount = payload.accountId.name.string
+        val genesis = "genesis"
```
Comment (arndey) [inline modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt:47]: create const

File: modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+import org.junit.jupiter.api.Test
+import kotlin.test.assertEquals
+
+class BlockStreamTest : IrohaTest<Iroha2Client>(account = ALICE_ACCOUNT_ID, keyPair = ALICE_KEYPAIR) {
+
+    @Test
+    @WithIroha([NewAccountWithMetadata::class])
+    fun `when subscribe to block stream then success`(): Unit = runBlocking {
+        var blocksResult = client.subscribeToBlockStream(1, 2)
+        val defaultDomain = "wonderland"
```
Comment (arndey) [inline modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt:34]: use const

File: modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt (hunk 3 above)
Comment (arndey) [reply modules/client/src/test/kotlin/jp/co/soramitsu/iroha2/BlockStreamTest.kt:64]: also I honesty don't understand the function name

**Review by arndey — CHANGES_REQUESTED** [review]
changes

File: modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/client/Iroha2Client.kt [hunk 10]
```diff
@@ -196,6 +201,39 @@ open class Iroha2Client(
         }
     }
 
+    /**
+     * Subscribe to block streaming
+     */
+    fun subscribeToBlockStream(startFromBlock: Long, blocksCount: Int): CompletableDeferred<List<VersionedBlockMessage>> {
```
[the remark this answers is not in the document]
Comment (Mingela) [reply modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/client/Iroha2Client.kt:207]: Please expand the issue description (or create new one) with the specific requirements, though that's exactly what I requested in the chat when the task came in 'to-do' 😢

**Review by Mingela — COMMENTED** [review]
Andrey's comments seem valid, I don't have more than those.

As for the API itself it'd great to return something like `Flux` (Reactor) or `Observable` (RxJava) or at least have a method expecting a logic to be applied to each block, i.e. `blockStream<T>(startFromBlock:Long, (VersionedBlockMessage) -> (T))`

**Review by Mingela — APPROVED** [review]

**Comment by Mingela** [comment]
Work to be continued within https://github.com/hyperledger/iroha-java/issues/330

**Pull request merged**